src/types.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * These represent the types in our abstract type system.
+ * We will normalize the raw XTP schema into these recursively defined types.
+ */
+
+export type XtpNormalizedKind =
+  'object' | 'enum' | 'map' | 'array' | 'string' |
+  'int32' | 'int64' | 'float' | 'double' |
+  'boolean' | 'date-time' | 'byte' | 'buffer'
+
+// applies type opts to a type on construction
+function cons(t: XtpNormalizedType, opts?: XtpTypeOpts): XtpNormalizedType {
+  // default them to false
+  t.nullable = (opts?.nullable === undefined) ? false : opts.nullable
+  t.required = (opts?.required === undefined) ? false : opts.required
+  return t
+}
+
+export interface XtpTypeOpts {
+  nullable?: boolean;
+  required?: boolean;
+}
+
+export interface XtpNormalizedType extends XtpTypeOpts {
+  kind: XtpNormalizedKind;
+}
+
+export class StringType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'string';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class Int32Type implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'int32';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class Int64Type implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'int64';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class FloatType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'float';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class DoubleType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'double';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class BooleanType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'boolean';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class ByteType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'byte';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class BufferType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'buffer';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class DateTimeType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'date-time';
+  constructor(opts?: XtpTypeOpts) {
+    cons(this, opts)
+  }
+}
+
+export class ObjectType implements XtpNormalizedType {
+  kind: XtpNormalizedKind = 'object';
+  name: string;
+  properties: Array<XtpNormalizedType>;
+
+  constructor(name: string, properties: Array<XtpNormalizedType>, opts?: XtpTypeOpts) {
+    this.name = name
+    this.properties = properties
+    cons(this, opts)
+  }
+}
+
+export class MapType implements XtpNormalizedType {
```

Comment:
`MapType` is raising some hairs – there's this concept of a semi-structured object in JSONSchema, where some properties are defined with specific types and other properties are allowed according to a pattern or a catch-all (`additionalProperties` or `patternProperties`). In Rust, at least, that's supported using `#[serde(flatten)]` – the additional properties can be flattened into a `HashMap<String, T>` on the parent object. Golang looks like it supports something similar with `map[string]*json.RawMessage`, though it requires some additional elbow grease.

I suppose my worry is that, while we can strictly disallow maps and structs now, if we codify `MapType` as distinct from an `ObjectType` it'll be hard to introduce that functionality later. (This kind of gets back to representing properties on an object as an array of `FieldDescriptor`)

Reply:
Map type should be used for typing what are now untyped objects. The distinction b/w an object and a map is dynamic keys, whereas object are for when you know the keys and types ahead of time.

A good example would be http headers i think. Today you can just say that it's an untyped object, but tomorrow you can say that it's a `Map<String, String>`.

Reply:
Ah yeah – the trick is that JSONSchema lets you represent both typed and dynamic keys in a single object. HTTP headers could be a useful example of a semi-structured object, too – you might say something like:

```yaml
ResponseHeaders:
  type: object
  properties:
    etag:
      type: string
  additionalProperties:
    type: string
```

and in rust you might produce:

```rust
#[derive(Deserialize)]
struct ResponseHeaders {
  etag: String

  #[serde(flatten)]
  additional_properties: HashMap<String, String>
}
```

or in TS:

```ts
interface ResponseHeaders {
  etag: string
  [additionalProperties: string]: string
}
```

(This not to say we need to enable this now, but if we're looking to track OpenAPI support it'd be useful to leave the door open to adding this down the line. I worry that adding a distinct `Map` type makes it harder to add support later, vs. pulling that info into the `FieldDescriptor`.)